src/multiversioning.zig

```diff
@@ -0,0 +1,892 @@
+                // is available.
+                unreachable;
+            }
+        };
+
+        const binary_offset = header.past.offsets[index];
+        const binary_size = header.past.sizes[index];
+        const binary_checksum = header.past.checksums[index];
+
+        // Our target release is physically embedded in the binary. Shuffle the bytes
+        // around, so that it's at the start, then truncate the descriptor so there's nothing
+        // trailing.
+        try os.lseek_SET(self.target_fd, self.target_body_offset.? + binary_offset);
+        const bytes_read = try os.read(self.target_fd, self.source_buffer[0..binary_size]);
+        assert(bytes_read == binary_size);
+
+        try os.lseek_SET(self.target_fd, 0);
+        const bytes_written = try os.write(self.target_fd, self.source_buffer[0..binary_size]);
+        assert(bytes_written == binary_size);
+
+        // Zero the remaining bytes in the file.
+        try os.ftruncate(self.target_fd, binary_size);
+
+        // Ensure the checksum matches the header. This could have been done above, but
+        // do it in a separate step to make sure.
+        const written_checksum = blk: {
+            try os.lseek_SET(self.target_fd, 0);
+            const bytes_read_for_checksum = try os.read(
+                self.target_fd,
+                self.source_buffer[0..binary_size],
+            );
+            assert(bytes_read_for_checksum == binary_size);
+            break :blk checksum.checksum(self.source_buffer[0..binary_size]);
+        };
+        assert(written_checksum == binary_checksum);
+
+        switch (builtin.target.os.tag) {
+            .linux => {
+                log.info("executing release {}...\n", .{release_target});
+
+                if (execveat(
+                    self.target_fd,
+                    "",
+                    self.args_envp.args,
+                    self.args_envp.envp,
+                    os.AT.EMPTY_PATH,
+                ) == 0) {
+                    unreachable;
+                } else {
+                    return error.ExecveatFailed;
+                }
+            },
+            else => @panic("exec_release unimplemented"),
+        }
+
+        unreachable;
+    }
+};
+
+pub fn self_exe_path(allocator: std.mem.Allocator) ![:0]const u8 {
+    var buf: [std.fs.MAX_PATH_BYTES]u8 = undefined;
+    const native_self_exe_path = try std.fs.selfExePath(&buf);
+
+    if (builtin.target.os.tag == .linux and std.mem.eql(
+        u8,
+        native_self_exe_path,
+        "/memfd:" ++ memfd_path ++ " (deleted)",
+    )) {
+        // Technically, "/memfd:tigerbeetle-multiversion (deleted)" is a valid path at which you
+        // could place your binary - please don't!
+        assert(std.fs.cwd().statFile(native_self_exe_path) catch null == null);
+
+        // Running from a memfd already; the real path is argv[0].
+        const path = std.mem.span(os.argv[0]);
```

Comment:
Should we duplicate this into the allocator? B/c right now I think we expect to caller to `allocator.free()` this.

Reply:
Great catch